File: src/content/algorithms/quick-sort/code.py
def quick_sort(array, low=0, high=None):  # #@function-def
    if high is None:
        high = len(array) - 1

    if low < high:  # #@base-case
        pivot_index = partition(array, low, high)  # #@partition
        quick_sort(array, low, pivot_index - 1)  # #@recursive-calls
        quick_sort(array, pivot_index + 1, high)  # #@recursive-calls

    return array


def partition(array, low, high):  # #@partition
    pivot = array[high]  # #@pivot

    i = low - 1

    for j in range(low, high):
        if array[j] <= pivot:  # #@comparison
            i += 1
            array[i], array[j] = array[j], array[i]  # #@swap

    array[i + 1], array[high] = array[high], array[i + 1]  # #@pivot-placement

    return i + 1

File: src/content/algorithms/quick-sort/code.rs
pub fn quick_sort<T: Ord>(arr: &mut [T]) {  // #@generic-function
    let len = arr.len();
    if len <= 1 {  // #@base-case
        return;
    }
    let pivot_index = partition(arr);  // #@slice-partition
    quick_sort(&mut arr[..pivot_index]);  // #@recursive-calls
    quick_sort(&mut arr[pivot_index + 1..]);  // #@recursive-calls
}

fn partition<T: Ord>(arr: &mut [T]) -> usize {  // #@slice-partition
    let len = arr.len();
    let pivot_index = len - 1;  // #@pivot
    let mut i = 0;

    for j in 0..pivot_index {
        if arr[j] <= arr[pivot_index] {  // #@comparison
            arr.swap(i, j);  // #@swap
            i += 1;
        }
    }

    arr.swap(i, pivot_index);  // #@pivot-placement
    i
}
